src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/LockFactory.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+
+namespace Microsoft.AspNetCore.Razor;
+
+internal class LockFactory
+{
+    // Specify recursion is supported, since an item with an upgradeable lock can still
+    // get another read lock on the same thread
+    protected readonly ReaderWriterLockSlim _lock = new(LockRecursionPolicy.SupportsRecursion);
+    private static readonly TimeSpan s_timeout = TimeSpan.FromMinutes(2);
```

Comment:
Is 2 minutes a bit long for production?

Reply:
Probably. Might need to make this something that isn't static and instead a method: 

```csharp
if (Debugger.IsAttached)
{
    return TimeSpan.Infinite; // if that's a thing
}

#if DEBUG 
return TimeSpan.FromMinutes(2);
#else 
return TimeSpan.FromSeconds(30);
#endif
```

Reply:
For debug-time, should it just be `TimeSpan.MaxValue`?